buffer/src/main/java/io/netty5/buffer/api/internal/Statics.java

```diff
@@ -393,4 +397,153 @@ static long nativeAddressOfDirectByteBuffer(ByteBuffer byteBuffer) {
+        long prefixes = maxFixes(needle, nl, m, needleStart, false);
+        int ell = Math.max((int) (suffixes >> 32), (int) (prefixes >> 32));
+        int period = Math.max((int) suffixes, (int) prefixes);
+        int length = Math.min(m - period, ell + 1);
+
+        if (equalsInner(needle, needleStart, needle, needleStart + period, length)) {
+            return bytesBeforeInnerPeriodic(haystack, hl, needle, nl, n, m, needleStart, haystackStart, ell, period);
+        }
+        return bytesBeforeInnerNonPeriodic(haystack, hl, needle, nl, n, m, needleStart, haystackStart, ell);
+    }
+
+    private static int bytesBeforeInnerPeriodic(Buffer haystack, UncheckedLoadByte hl,
+                                                Buffer needle, UncheckedLoadByte nl,
+                                                int n, int m, int needleStart, int haystackStart, int ell, int period) {
+        int i;
+        int j = 0;
+        int memory = -1;
+        while (j <= n - m) {
+            i = Math.max(ell, memory) + 1;
+            while (i < m && nl.load(needle, i + needleStart) == hl.load(haystack, i + j + haystackStart)) {
+                ++i;
+            }
+            if (i > n) {
+                return -1;
+            }
+            if (i >= m) {
+                i = ell;
+                while (i > memory && nl.load(needle, i + needleStart) == hl.load(haystack, i + j + haystackStart)) {
+                    --i;
+                }
+                if (i <= memory) {
+                    return j;
+                }
+                j += period;
+                memory = m - period - 1;
+            } else {
+                j += i - ell;
+                memory = -1;
+            }
+        }
+        return -1;
+    }
+
+    private static int bytesBeforeInnerNonPeriodic(Buffer haystack, UncheckedLoadByte hl,
+                                                   Buffer needle, UncheckedLoadByte nl,
+                                                   int n, int m, int needleStart, int haystackStart, int ell) {
+        int i;
+        int j = 0;
+        int period = Math.max(ell + 1, m - ell - 1) + 1;
+        while (j <= n - m) {
+            i = ell + 1;
+            while (i < m && nl.load(needle, i + needleStart) == hl.load(haystack, i + j + haystackStart)) {
+                ++i;
+            }
+            if (i > n) {
+                return -1;
+            }
+            if (i >= m) {
+                i = ell;
+                while (i >= 0 && nl.load(needle, i + needleStart) == hl.load(haystack, i + j + haystackStart)) {
+                    --i;
+                }
+                if (i < 0) {
+                    return j;
+                }
+                j += period;
+            } else {
+                j += i - ell;
+            }
+        }
+        return -1;
+    }
+
+    private static long maxFixes(Buffer needle, UncheckedLoadByte nl, int m, int start, boolean isSuffix) {
```

Comment:
Can we make this method easier to understand from a review point of view? I mean, those `int p, ms, j, k` does not provide much context.

Reply:
Those variable names are taken from http://www-igm.univ-mlv.fr/~lecroq/string/node26.html - some of them I'm not able to figure out what to call them.

Reply:
Go some hints at the names from the original paper. The `k` variable isn't given a better name, though.